src/solidlsp/ls.py

```diff
@@ -1566,56 +1676,110 @@ def request_defining_symbol(
 
         return defining_symbol
 
-    @property
-    def cache_path(self) -> Path:
-        """
-        The path to the cache file for the document symbols.
-        """
-        return (
-            Path(self.repository_root_path)
-            / self._solidlsp_settings.project_data_relative_path
-            / self.CACHE_FOLDER_NAME
-            / self.language_id
-            / "document_symbols_cache.pkl"
-        )
+    def _save_raw_document_symbols_cache(self):
+        cache_file = self.cache_dir / self.RAW_DOCUMENT_SYMBOL_CACHE_FILENAME
 
-    def save_cache(self):
-        with self._cache_lock:
-            if not self._cache_has_changed:
-                self.logger.log("No changes to document symbols cache, skipping save", logging.DEBUG)
-                return
+        if not self._raw_document_symbols_cache_is_modified:
+            log.debug("No changes to raw document symbols cache, skipping save")
+            return
+
+        log.info("Saving updated raw document symbols cache to %s", cache_file)
+        try:
+            save_cache(str(cache_file), self._raw_document_symbols_cache_version(), self._raw_document_symbols_cache)
+            self._raw_document_symbols_cache_is_modified = False
+        except Exception as e:
+            log.error(
+                "Failed to save raw document symbols cache to %s: %s. Note: this may have resulted in a corrupted cache file.",
+                cache_file,
+                e,
+            )
+
+    def _raw_document_symbols_cache_version(self):
+        return (self.RAW_DOCUMENT_SYMBOLS_CACHE_VERSION, self._ls_specific_raw_document_symbols_cache_version)
+
+    def _load_raw_document_symbols_cache(self):
+        cache_file = self.cache_dir / self.RAW_DOCUMENT_SYMBOL_CACHE_FILENAME
 
-            self.logger.log(f"Saving updated document symbols cache to {self.cache_path}", logging.INFO)
-            self.cache_path.parent.mkdir(parents=True, exist_ok=True)
+        if not cache_file.exists():
+            # check for legacy cache to load to migrate
+            legacy_cache_file = self.cache_dir / self.RAW_DOCUMENT_SYMBOL_CACHE_FILENAME_LEGACY_FALLBACK
+            if legacy_cache_file.exists():
+                try:
+                    legacy_cache: dict[
+                        str, tuple[str, tuple[list[ls_types.UnifiedSymbolInformation], list[ls_types.UnifiedSymbolInformation]]]
+                    ] = load_pickle(legacy_cache_file)
+                    log.info("Migrating legacy document symbols cache with %d entries", len(legacy_cache))
```

Comment:
Maybe on warning level? I also think we should delete the legacy cache file and create a new one, so I propose adding a separate method `_migrate_legacy_cache` that takes care of that, so loading looks like

```python
def _load_raw_document_symbols_cache(self):
    if not cache_file.exists() and legacy_file_cache.exists():
        self._migrate_legacy_cache()
    if cache_file_exists():
         ...
```

Reply:
Why warning level? Warnings are reserved for something the user should probably pay attention to, because there might be a problem. This is just a normal event, so warning level would be inappropriate.

We can think about doing the deletion once we are done completely with this topic. Right now, keeping it around is better when switching between branches and testing the migrations.

Reply:
Ok, then deletion at 1.0.0 release

Reply:
Yes. I am merging this branch.